### PR TITLE
Move AT_EXECFN to fallback for /proc/self/exe

### LINKED ISSUE #78941: dotnet EXE application doesn't work with shebang on Linux
### Description

When building a self-contained console application targeting .NET 7, the console application doesn't work with shebang in Linux.

However, when building the same console application targeting .NET 6, the produced console application works fine with shebang.

This causes powershell v7.3.0 to stop working with shebang on Linux. See https://github.com/PowerShell/PowerShell/issues/18510

### Build a simple self-contained console application targeting .NET 7

The console application is very simple:
```c#
using System;

namespace MyApp
{
    internal class Program
    {
        static void Main(string[] args)
        {
            for (int i = 0; i < args.Length; i++)
            {
                Console.WriteLine($"<Arg {i}>: {args[i]}");
            }
        }
    }
}
```

The .csproj file:
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
  </PropertyGroup>

</Project>
```

Build a self-contained console application from the C# code, targeting .NET 7:
```
dotnet publish -r linux-x64 -c Release --self-contained
```

### Create the script file that uses shebang to run the console application

The content of the script file `dotnet7`:
```
#!/home/daxian/apps/1self/bin/Release/net7.0/linux-x64/publish/1self
$true
```

Then, run `chmod` to make the file an executable
```
chmod +x ./dotnet7
```

At last, run `dotnet7` in bash:
```
./dotnet7
```

### Expected behavior

The console application should run and print the arguments it receives.

When building against .NET 6, it works as expected, and here is the screenshot of the result:

image

### Actual behavior

When building against .NET 7, it stops working with the following error message:

```
bash:35> ./dotnet7
The application to execute does not exist: '/home/daxian/temp/1self.dll'.
```

image

### Regression?

Yes, it's a regression from .NET 6

### Known Workarounds

No known workaround so far.

### Configuration

> Which version of .NET is the code running on?

.NET 7 -- SDK version 7.0.100, 

> What OS and version, and what distro if applicable?

WSLv2 - Ubuntu 18.04

> What is the architecture (x64, x86, ARM, ARM64)?

I reproduced the issue on x64, but didn't try other architectures.

### Other information

_No response_

### PR DESCRIPTION
**Pull request by am11**
In https://github.com/dotnet/runtime/pull/66874, the missing header was added which enabled the fast (aux vector) path on Linux.

The issue with `AT_EXECFN` is that it returns the path which was passed to the process via [`namei`](https://github.com/torvalds/linux/blob/master/fs/namei.c). This breaks the scenario where the .NET process is used as interpreter (namely PowerShell's `#!/bin/pwsh`) and returns the path of script containing the shebang link to the interpreter rather than the interpreter's path. In other words, `AT_EXCFN` implementation != `/proc/self/exe`. In all the usages of `minipal_getexepath()` in runtime repo, we need the behavior of latter on Linux.

<del>The fix is to remove the usage of `AT_EXCFN` and rely solely on `/proc/self/exe` for Linux (which we were accidentally using in .NET 6 due to the missing header).</del>

The fix is to flip the order so `AT_EXCFN` is used as a fallback to `/proc/self/exe`.

Fixes #78941 (we will probably need to backport this to .NET 7)

### CODE REVIEW

**Comment by janvorli** [comment]
This is unfortunate w.r.t. asks on minimizing usage of the `/proc` file system (one example is #2534) or issues where `/proc/self/exe` doesn't actually return the "real" exe path (#47280). I wonder if there is a way to handle the case this PR is fixing differently.

**Comment by janvorli** [comment]
I have looked at the `getauxval` doc and figured out a solution that doesn't use the `/proc/self/exe` and yet doesn't have the problem this issue is fixing:

```C
unsigned long entry = getauxval(AT_ENTRY);
    st = dladdr((void*)entry, &info2);

    if (st != 0)
    {
        printf("From AT_ENTRY: %s\n", info2.dli_fname);
    }
```
I've created a .sh script with the shebang pointing to my test app with the code above and when I ran it it printed this:
```
From AT_ENTRY: /home/janvorli/test/testatexecfn/test
```

**Comment by am11** [comment]
> This is unfortunate w.r.t. asks on minimizing usage of the `/proc` file system 

Yes, that's how we discovered the missing header https://github.com/dotnet/runtime/pull/66874#issuecomment-1072956444 when running a .NET app on a system with broken procfs. After that I sent other PRs to remove the usage of procfs in runtime and sdk repos.

This is a correctness fix ; both `#if HAVE_GETAUXVAL` and `#else` branches should be equivalent, but they are not. `/proc/self/exe` is not perfect, but it just works on a healthy system. This is a low risk fix because we are reverting to .NET 6 behavior.

> AT_ENTRY

Could you point me where in the docs does it mention the AT_ENTRY based  solution to get the entry? I have not found any references of anyone using this approach in production. It is either /proc/self/exe or AT_EXCFN. This thread has some discussion on the pros and cons https://lkml.iu.edu/hypermail/linux/kernel/0808.1/3101.html. Each method comes with its own caveats, so I am not confident to jump right on AT_ENTRY solution (which we have never tested before) for backport fix.

**Comment by am11** [comment]
> This is unfortunate w.r.t. asks on minimizing usage of the `/proc` file system 

Yes, that's how we discovered the missing header https://github.com/dotnet/runtime/pull/66874#issuecomment-1072956444 when running a .NET app on a system with broken procfs. After that I sent other PRs to remove the usage of procfs in runtime and sdk repos.

This PR, however, is a correctness fix; both `#if HAVE_GETAUXVAL` and `#else` branches should be equivalent, but they are not. `/proc/self/exe` is not perfect, but it just works on a healthy system. This is a low risk fix because we are reverting to .NET 6 behavior.

> AT_ENTRY

Could you point me where the docs mention the `AT_ENTRY` based solution to get the executable path? I have not found any references where someone is using this approach in production. It is either `/proc/self/exe` or `AT_EXCFN`. This thread has some discussion on the pros and cons https://lkml.iu.edu/hypermail/linux/kernel/0808.1/3101.html. Each method comes with its own caveats, so I am not confident to jump right on `AT_ENTRY` solution (which we have never tested before) for backport fix.

**Comment by janvorli** [comment]
@am11 I've found AT_ENTRY documented in the `getauxval` doc. It is the entry point address, so it must be in the actual executable. So I got the idea of using `dladdr` that can get the full path to the executable. I have not tried to search for it anywhere, so it is possible that no one is using that.
It is possible that there are some pitfalls, so I would definitely not jump into backporting it until we were sure that it works reliably.

I have actually got an additional idea on fixing this issue. What if we just flipped the order of reading the `/proc/self/exe` and the `AT_EXECFN` instead of getting rid of the latter? Then the `AT_EXECFN` would be a fallback for cases when the `/proc/self/exe` isn't there.

**Comment by am11** [comment]
> What if we just flipped the order of reading the `/proc/self/exe` and the `AT_EXECFN` instead of getting rid of the latter? Then the `AT_EXECFN` would be a fallback for cases when the `/proc/self/exe` isn't there.

I like this idea, thank you. I think it will minimize the issue. It will only affect the system with broken procfs and using .NET process as interpreter; which is likely a very small subset of users who can find a way to fix their procfs.

File: src/native/minipal/getexepath.h [hunk 1]
```diff
@@ -75,21 +75,28 @@ static inline char* minipal_getexepath(void)
...
+
+#if HAVE_GETAUXVAL && defined(AT_EXECFN)
+    // fallback to AT_EXECFN, which does not work properly in rare cases
+    // when .NET process is set as interpreter (shebang).
+    const char* exePath = (const char *)(getauxval(AT_EXECFN));
+    if (exePath && !errno)
+    {
+        return realpath(exePath, NULL);
+    }
+#endif // HAVE_GETAUXVAL && defined(AT_EXECFN)
```
Comment (janvorli) [inline src/native/minipal/getexepath.h:99]: This function is missing return in case everything fails.

Comment (am11) [reply src/native/minipal/getexepath.h:99]:
```suggestion
#endif // HAVE_GETAUXVAL && defined(AT_EXECFN)

    return NULL;
```

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Comment by janvorli** [comment]
All the CI failures are known issues unrelated to this change.

**Comment by am11** [comment]
Thank you.

I think we should backport this to .NET 7 (to unblock PowerShell users). cc @agocke, @vitek-karas

**Pull request merged**